neuralmonkey/decoders/sequence_decoder.py

```diff
@@ -0,0 +1,258 @@
+    # pylint: enable=protected-access
+    return cast(Type, NamedTuple(name, ext_fields))
+
+
+# The LoopState is a structure that works with the tf.while_loop function
+# the decoder loop state stores all the information that is not invariant
+# for the decoder run.
+# pylint: disable=invalid-name
+DecoderHistories = NamedTuple(
+    "DecoderHistories",
+    [("logits", tf.TensorArray),
+     ("decoder_outputs", tf.TensorArray),
+     ("mask", tf.TensorArray)])  # float matrix, 0s and 1s
+
+DecoderConstants = NamedTuple(
+    "DecoderConstants",
+    [("train_inputs", Optional[tf.Tensor])])
+
+DecoderFeedables = NamedTuple(
+    "DecoderFeedables",
+    [("step", tf.Tensor),  # 1D int, number of the step
+     ("finished", tf.Tensor),  # batch-sized, bool
+     ("input_symbol", tf.Tensor),
+     ("prev_logits", tf.Tensor)])
+
+LoopState = NamedTuple(
+    "LoopState",
+    [("histories", Any),
+     ("constants", Any),
+     ("feedables", Any)])
+# pylint: enable=invalid-name
+
+
+# pylint: disable=too-many-public-methods
+# TODO More refactoring needed.. (?)
+class SequenceDecoder(ModelPart):
+
+    def __init__(self,
+                 name: str,
+                 vocabulary: Vocabulary,
+                 data_id: str,
+                 max_output_len: int,
+                 dropout_keep_prob: float = 1.0,
+                 save_checkpoint: str = None,
+                 load_checkpoint: str = None) -> None:
+        ModelPart.__init__(self, name, save_checkpoint, load_checkpoint)
+        log("Initializing decoder, name: '{}'".format(name))
+
+        self.vocabulary = vocabulary
+        self.data_id = data_id
+        self.max_output_len = max_output_len
+        self.dropout_keep_prob = dropout_keep_prob
+
+        # TODO check the values of the parameters (vocab, ...)
+
+        with self.use_scope():
+            self.train_mode = tf.placeholder(tf.bool, [], "train_mode")
+            self.go_symbols = tf.placeholder(tf.int32, [None], "go_symbols")
+
+            self.train_inputs = tf.placeholder(
+                tf.int32, [None, None], "train_inputs")
+            self.train_mask = tf.placeholder(
+                tf.float32, [None, None], "train_mask")
+
+    @tensor
+    def batch_size(self) -> tf.Tensor:
+        return tf.shape(self.go_symbols)[0]
+
+    @tensor
+    def decoding_w(self) -> tf.Variable:
+        with tf.name_scope("output_projection"):
+            return tf.get_variable(
+                "logit_matrix",
+                [self.output_dimension, len(self.vocabulary)],
+                initializer=tf.random_uniform_initializer(-0.5, 0.5))
+
+    @tensor
+    def decoding_b(self) -> tf.Variable:
+        with tf.name_scope("output_projection"):
+            return tf.get_variable(
+                "logit_bias", [len(self.vocabulary)],
+                initializer=tf.constant_initializer(
+                    - math.log(len(self.vocabulary))))
+
+    def get_logits(self, state: tf.Tensor) -> tf.Tensor:
```

Comment:
docstring:
Project the output layer of the decoder to logits over the vocabulary.

Reply:
DONE